### PR TITLE
Split GetPortsAndProtocol function to 4 separate functions

### PR DESCRIPTION
**Pull request by panslava**
Return values are independent, and there is no reason to do it in single function

### CODE REVIEW

File: pkg/loadbalancers/forwarding_rules.go [hunk 1]
```diff
@@ -242,7 +242,8 @@ func (l *L4) ensureForwardingRule(loadBalancerName, bsLink string, options gce.I
 		}()
 	}
 
-	ports, _, _, protocol := utils.GetPortsAndProtocol(l.Service.Spec.Ports)
+	servicePorts := l.Service.Spec.Ports
+	ports, protocol := utils.GetPorts(servicePorts), utils.GetProtocol(servicePorts)
```
Comment (cezarygerard) [inline pkg/loadbalancers/forwarding_rules.go:246]: please split into 2 lines

File: pkg/loadbalancers/l4.go [hunk 2]
```diff
@@ -213,7 +213,8 @@ func (l *L4) EnsureInternalLoadBalancer(nodeNames []string, svc *corev1.Service)
 	}
 	result.Annotations[annotations.HealthcheckKey] = hcResult.HCName
 
-	_, portRanges, _, protocol := utils.GetPortsAndProtocol(l.Service.Spec.Ports)
+	servicePorts := l.Service.Spec.Ports
+	portRanges, protocol := utils.GetServicePortRanges(servicePorts), utils.GetProtocol(servicePorts)
```
Comment (cezarygerard) [inline pkg/loadbalancers/l4.go:217]: please split into 2 lines

File: pkg/loadbalancers/l4netlb.go [hunk 3]
```diff
@@ -131,7 +131,8 @@ func (l4netlb *L4NetLB) EnsureFrontend(nodeNames []string, svc *corev1.Service)
 	result.Annotations[annotations.HealthcheckKey] = hcResult.HCName
 
 	name := l4netlb.ServicePort.BackendName()
-	_, portRanges, _, protocol := utils.GetPortsAndProtocol(l4netlb.Service.Spec.Ports)
+	servicePorts := l4netlb.Service.Spec.Ports
+	portRanges, protocol := utils.GetServicePortRanges(servicePorts), utils.GetProtocol(servicePorts)
```
Comment (cezarygerard) [inline pkg/loadbalancers/l4netlb.go:135]: please split into 2 lines

**Comment by cezarygerard** [comment]
small comments, otherwise LGTM

**Review by cezarygerard — CHANGES_REQUESTED** [review]

**Comment by panslava** [comment]
I also think I should add Unit tests here, cause these functions were not covered

File: pkg/loadbalancers/forwarding_rules.go (hunk 1 above)
Comment (panslava) [reply pkg/loadbalancers/forwarding_rules.go]: Done

File: pkg/loadbalancers/l4.go (hunk 2 above)
Comment (panslava) [reply pkg/loadbalancers/l4.go]: Done

File: pkg/loadbalancers/l4netlb.go (hunk 3 above)
Comment (panslava) [reply pkg/loadbalancers/l4netlb.go]: Done

File: pkg/utils/utils_test.go [hunk 4]
```diff
@@ -1326,39 +1326,275 @@ func TestIsLoadBalancerType(t *testing.T) {
...
+			nodePorts := GetNodePorts(tc.ports)
+
+			if !reflect.DeepEqual(nodePorts, tc.expectedNodePorts) {
+				t.Errorf("GetNodePorts returned %v, not equal to expected node ports = %v", nodePorts, tc.expectedNodePorts)
+			}
+		})
+	}
+}
+
+func TestGetPorts(t *testing.T) {
```
Comment (cezarygerard) [inline pkg/utils/utils_test.go:1417]: this does not need to be a table test
but I can live with it ;-)

File: pkg/utils/utils_test.go [hunk 5]
```diff
@@ -1326,39 +1326,275 @@ func TestIsLoadBalancerType(t *testing.T) {
...
+	testCases := []struct {
```
Comment (cezarygerard) [inline pkg/utils/utils_test.go:1418]: this formatting wll read better:

```
testCases := []struct {
		ports         []api_v1.ServicePort
		expectedPorts []string
		desc          string
	}{
		{
			ports:         []api_v1.ServicePort{},
			expectedPorts: []string{},
			desc:          "Empty ports should return empty slice",
		},
		{
			ports: []api_v1.ServicePort{
				{Port: 80}, {Port: 81}, {Port: 3000},
			},
			expectedPorts: []string{"80", "81", "3000"},
			desc:          "Multiple ports",
		},
	}
```

File: pkg/utils/utils_test.go [hunk 6]
```diff
@@ -1326,39 +1326,275 @@ func TestIsLoadBalancerType(t *testing.T) {
...
+	}
+}
+
+func TestGetServicePortRanges(t *testing.T) {
+	testCases := []struct {
+		ports          []api_v1.ServicePort
+		expectedRanges []string
+		desc           string
+	}{
+		{desc: "All Unique", ports: []api_v1.ServicePort{{
```
Comment (cezarygerard) [inline pkg/utils/utils_test.go:1462]: this is some weird formatting, try this:

```
		{
			desc: "All Unique",
			ports: []api_v1.ServicePort{
				{Port: 8}, {Port: 66}, {Port: 23}, {Port: 13}, {Port: 89},
			},
			expectedRanges: []string{"8", "13", "23", "66", "89"},
		},
```
Some for other test cases

File: pkg/utils/utils_test.go [hunk 7]
```diff
@@ -1326,39 +1326,275 @@ func TestIsLoadBalancerType(t *testing.T) {
...
+			protocol := GetProtocol(tc.ports)
+
+			if protocol != tc.expectedProtocol {
+				t.Errorf("GetProtocol returned %v, not equal to expected protocol = %v", protocol, tc.expectedProtocol)
+			}
+		})
+	}
+}
+
+func TestGetNodePorts(t *testing.T) {
```
Comment (cezarygerard) [inline pkg/utils/utils_test.go:1378]: Check my comments in TestGetPorts
1) this does not need to be a table test
2) inlining some structs in the test case may help readability

**Review by cezarygerard — COMMENTED** [review]
just readability concerns

again ;-)

File: pkg/utils/utils_test.go (hunk 4 above)
Comment (panslava) [reply pkg/utils/utils_test.go:1409]: I think there is some value in checking it for "edge" condition (empty slice) and with some real value

File: pkg/utils/utils_test.go (hunk 5 above)
Comment (panslava) [reply pkg/utils/utils_test.go:1410]: Fixed

File: pkg/utils/utils_test.go (hunk 7 above)
Comment (panslava) [reply pkg/utils/utils_test.go:1378]: Fixed formatting

File: pkg/utils/utils_test.go (hunk 6 above)
Comment (panslava) [reply pkg/utils/utils_test.go]: Fixed

File: pkg/utils/utils_test.go (hunk 4 above)
Comment (cezarygerard) [reply pkg/utils/utils_test.go:1409]: yeah, I did not notice that the test actually has more than 1 condition.

**Pull request merged**